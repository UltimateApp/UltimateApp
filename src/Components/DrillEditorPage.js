import React, { useState } from 'react';
import { StyleSheet, View, Text, ScrollView, KeyboardAvoidingView, Platform } from 'react-native';
import { connect } from 'react-redux';
import { Formik, FieldArray } from 'formik';
import * as Yup from 'yup';

import { DrillTypes, FrisbeeGoals } from '../Fixtures/config';
import theme from '../styles/theme.style';
import I18n from '../utils/i18n';
import { generateUuid } from '../utils/uuid';
import { saveDrill } from '../Store/Actions/drillAction';
import Button from './shared/Button';
import Input from './shared/form/Input';

// TODO: Stocker correctement les champs number
// TODO: Gérer différemment le partage de drill de l'app et de drill custom (pour l'instant empêcher ?)
// TODO: Pouvoir créer une animation
// TODO: Écrire des tests plus précis du formulaire
// TODO: Meilleures validations (notamment sur steps)

const newStep = {
  id: 0,
  title: '',
  animation: {
    positions: [[], []],
    ids: [],
    texts: [],
    background: 'endzone',
  },
  vimeoId: undefined,
  youtube: undefined,
  instruction: '',
};

const newDrill = {
  id: generateUuid(),
  custom: true,
  type: DrillTypes.FRISBEE,
  visibleInList: true,
  author: 'Martin',
  title: 'Mon drill',
  image: undefined,
  description: 'Ma super description',
  minimalPlayersNumber: 8, // Should be a string, then casted back to integer before save?
  inGame: 'ça sert un peu',
  equipmentLabel: undefined,
  equipment: '1 disque',
  durationInMinutes: undefined,
  intensity: undefined,
  goals: [],
  seasonTiming: undefined,
  level: undefined,
  steps: [newStep],
};

export const DrillEditorPage = (props) => {
  const [currentDrill, setCurrentDrill] = useState(newDrill);

  const validationSchema = Yup.object({
    title: Yup.string()
      .trim()
      .required(I18n.t('editor.playTitle.empty'))
      .notOneOf(
        props.customDrills.map((drill) => drill.title),
        I18n.t('editor.playTitle.alreadyExists'),
      ),
    description: Yup.string().trim(),
    author: Yup.string().trim(),
    minimalPlayersNumber: Yup.number().positive(),
    equipment: Yup.string().trim(),
  });

  const behavior = Platform.select({
    ios: 'padding',
    android: 'height',
  });

  return (
    <KeyboardAvoidingView style={styles.drillEditorPage} behavior={behavior}>
      <ScrollView>
        <Formik
          initialValues={currentDrill}
          validationSchema={validationSchema}
          onSubmit={(values) => {
            const newCurrentDrill = { ...currentDrill, ...values };
            props.saveDrill(newCurrentDrill);
            props.navigation.navigate('DrillListPage', { type: DrillTypes.FRISBEE });
          }}
        >
          {({ handleSubmit, handleChange, errors, values, touched, isValid }) => (
            <View>
              <Input fieldName="author" label="Author" />
              <Input fieldName="title" label="Title" />
              <Input fieldName="image" label="URL de l'image" />
              <Input fieldName="description" label="Description" />
              <Input fieldName="minimalPlayersNumber" keyboardType="number-pad" label="Minimal number of players" />
              <Input fieldName="inGame" label="Utilité en jeu" />
              <Input fieldName="equipement" label="Equipment" />
              <Input fieldName="duration" keyboardType="number-pad" label="Duration" />
              <FieldArray
                name="steps"
                render={({ push, remove }) => (
                  <View>
                    <View style={styles.stepHeader}>
                      <Text style={styles.stepHeadertext}>Steps</Text>
                      <Button onPress={() => push({ ...newStep, id: values.steps.length })} text="+" small light />
                    </View>
                    {values.steps.map((step, index) => (
                      <View style={styles.step}>
                        <Text>Step {index + 1}</Text>
                        <Input fieldName={`steps[${index}].title`} label="Title" />
                        <Input fieldName={`steps[${index}].instruction`} label="Instructions" />
                        <Input fieldName={`steps[${index}].vimeoId`} label="Identifiant Vimeo" />
                        <Input fieldName={`steps[${index}].youtube`} label="URL Youtube" />
                        <Button onPress={() => remove(index)} text="-" small light style={styles.button} />
                      </View>
                    ))}
                  </View>
                )}
              />
              <Button onPress={handleSubmit} text={I18n.t('drillEditorPage.cta')} style={styles.cta} />
            </View>
          )}
        </Formik>
      </ScrollView>
    </KeyboardAvoidingView>
  );
};

const mapStateToProps = (state) => {
  return {
    customDrills: state.customDrills,
  };
};

const mapDispatchToProps = { saveDrill };

export default connect(mapStateToProps, mapDispatchToProps)(DrillEditorPage);

const styles = StyleSheet.create({
  drillEditorPage: {
    flex: 1,
    backgroundColor: theme.BACKGROUND_COLOR_LIGHT,
    padding: 10,
  },
  step: {
    paddingBottom: 10,
    marginLeft: 10,
    marginBottom: 10,
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: theme.BORDER_COLOR_BUTTON,
  },
  stepHeader: {
    flexDirection: 'row',
  },
  stepHeadertext: {
    fontSize: theme.FONT_SIZE_LARGE,
    marginRight: 10,
  },
  button: {
    alignSelf: 'flex-start',
  },
});
