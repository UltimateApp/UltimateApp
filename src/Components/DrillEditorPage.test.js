import React from 'react';
import { create, act } from 'react-test-renderer';

import DrillEditorPage from './DrillEditorPage';

describe('<DrillEditorPage />', () => {
  it('renders correctly', async () => {
    const tree = create(<DrillEditorPage />).toJSON();
    await act(async () => expect(tree).toMatchSnapshot());
  });
});
