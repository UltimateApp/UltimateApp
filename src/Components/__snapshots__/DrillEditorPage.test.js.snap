// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DrillEditorPage /> renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "padding": 10,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTScrollView>
    <View>
      <View>
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 5,
              }
            }
          >
            Author
          </Text>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#e5e5e5",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "padding": 5,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              rejectResponderTermination={true}
              underlineColorAndroid="transparent"
              value="Martin"
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 5,
              }
            }
          >
            Title
          </Text>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#e5e5e5",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "padding": 5,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              rejectResponderTermination={true}
              underlineColorAndroid="transparent"
              value="Mon drill"
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 5,
              }
            }
          >
            URL de l'image
          </Text>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#e5e5e5",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "padding": 5,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              rejectResponderTermination={true}
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 5,
              }
            }
          >
            Description
          </Text>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#e5e5e5",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "padding": 5,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              rejectResponderTermination={true}
              underlineColorAndroid="transparent"
              value="Ma super description"
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 5,
              }
            }
          >
            Minimal number of players
          </Text>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#e5e5e5",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "padding": 5,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              keyboardType="number-pad"
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              rejectResponderTermination={true}
              underlineColorAndroid="transparent"
              value={8}
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 5,
              }
            }
          >
            Utilité en jeu
          </Text>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#e5e5e5",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "padding": 5,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              rejectResponderTermination={true}
              underlineColorAndroid="transparent"
              value="ça sert un peu"
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 5,
              }
            }
          >
            Equipment
          </Text>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#e5e5e5",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "padding": 5,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              rejectResponderTermination={true}
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
        <View
          style={
            Object {
              "marginBottom": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "marginBottom": 5,
              }
            }
          >
            Duration
          </Text>
          <View
            style={
              Array [
                Object {
                  "borderColor": "#e5e5e5",
                  "borderRadius": 5,
                  "borderWidth": 0.5,
                  "padding": 5,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              keyboardType="number-pad"
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              rejectResponderTermination={true}
              underlineColorAndroid="transparent"
            />
          </View>
        </View>
        <View>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 20,
                  "marginRight": 10,
                }
              }
            >
              Steps
            </Text>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "backgroundColor": "#fff",
                  "borderColor": "#2a9e91",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 30,
                  "justifyContent": "center",
                  "opacity": 1,
                  "paddingHorizontal": 10,
                  "paddingVertical": 0,
                  "width": "auto",
                }
              }
              testID="button"
            >
              <Text
                style={
                  Object {
                    "color": "#2a9e91",
                    "fontSize": 16,
                    "fontWeight": "bold",
                  }
                }
              >
                +
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "borderBottomColor": "#e5e5e5",
                "borderBottomWidth": 0.5,
                "marginBottom": 10,
                "marginLeft": 10,
                "paddingBottom": 10,
              }
            }
          >
            <Text>
              Step 
              1
            </Text>
            <View
              style={
                Object {
                  "marginBottom": 10,
                }
              }
            >
              <Text
                style={
                  Object {
                    "marginBottom": 5,
                  }
                }
              >
                Title
              </Text>
              <View
                style={
                  Array [
                    Object {
                      "borderColor": "#e5e5e5",
                      "borderRadius": 5,
                      "borderWidth": 0.5,
                      "padding": 5,
                    },
                    undefined,
                  ]
                }
              >
                <TextInput
                  allowFontScaling={true}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  rejectResponderTermination={true}
                  underlineColorAndroid="transparent"
                  value=""
                />
              </View>
            </View>
            <View
              style={
                Object {
                  "marginBottom": 10,
                }
              }
            >
              <Text
                style={
                  Object {
                    "marginBottom": 5,
                  }
                }
              >
                Instructions
              </Text>
              <View
                style={
                  Array [
                    Object {
                      "borderColor": "#e5e5e5",
                      "borderRadius": 5,
                      "borderWidth": 0.5,
                      "padding": 5,
                    },
                    undefined,
                  ]
                }
              >
                <TextInput
                  allowFontScaling={true}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  rejectResponderTermination={true}
                  underlineColorAndroid="transparent"
                  value=""
                />
              </View>
            </View>
            <View
              style={
                Object {
                  "marginBottom": 10,
                }
              }
            >
              <Text
                style={
                  Object {
                    "marginBottom": 5,
                  }
                }
              >
                Identifiant Vimeo
              </Text>
              <View
                style={
                  Array [
                    Object {
                      "borderColor": "#e5e5e5",
                      "borderRadius": 5,
                      "borderWidth": 0.5,
                      "padding": 5,
                    },
                    undefined,
                  ]
                }
              >
                <TextInput
                  allowFontScaling={true}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  rejectResponderTermination={true}
                  underlineColorAndroid="transparent"
                />
              </View>
            </View>
            <View
              style={
                Object {
                  "marginBottom": 10,
                }
              }
            >
              <Text
                style={
                  Object {
                    "marginBottom": 5,
                  }
                }
              >
                URL Youtube
              </Text>
              <View
                style={
                  Array [
                    Object {
                      "borderColor": "#e5e5e5",
                      "borderRadius": 5,
                      "borderWidth": 0.5,
                      "padding": 5,
                    },
                    undefined,
                  ]
                }
              >
                <TextInput
                  allowFontScaling={true}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  rejectResponderTermination={true}
                  underlineColorAndroid="transparent"
                />
              </View>
            </View>
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "flex-start",
                  "backgroundColor": "#fff",
                  "borderColor": "#2a9e91",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 30,
                  "justifyContent": "center",
                  "opacity": 1,
                  "paddingHorizontal": 10,
                  "paddingVertical": 0,
                  "width": "auto",
                }
              }
              testID="button"
            >
              <Text
                style={
                  Object {
                    "color": "#2a9e91",
                    "fontSize": 16,
                    "fontWeight": "bold",
                  }
                }
              >
                -
              </Text>
            </View>
          </View>
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#2a9e91",
              "borderRadius": 5,
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "center",
              "opacity": 1,
              "paddingHorizontal": 20,
              "paddingVertical": 10,
              "width": "80%",
            }
          }
          testID="button"
        >
          <Text
            style={
              Object {
                "color": "#fff",
                "fontSize": 16,
                "textAlign": "center",
              }
            }
          >
            Save
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
